conn/conn.go

```diff
@@ -0,0 +1,131 @@
+}
+
+// Read calls the underlying connection's Read method.
+func (rlc *RLConnection) Read(b []byte) (n int, err error) {
+	return rlc.conn.Read(b)
+}
+
+// SetLimit sets the rate limit of the connection.
+func (rlc *RLConnection) SetLimit(packetSize uint64, packetsPerSecond uint64) {
+	rlc.packetSize = packetSize
+	rlc.packetsPerSecond = packetsPerSecond
+	rlc.lastWrite = time.Now()
+	rlc.writtenBytes = rlc.packetSize
+}
+
+// Write writes data to the underlying connection without exceeding the rate
+// limit.
+func (rlc *RLConnection) Write(b []byte) (n int, err error) {
+	// If there is no rate limit, we can write everything at once.
+	if rlc.packetsPerSecond == 0 {
+		return rlc.conn.Write(b)
+	}
+
+	// Start a feeder thread that feeds us the data at the right speed.
+	packets := make(chan []byte)
+	go func() {
+		defer close(packets)
+		if rlc.writtenBytes < rlc.packetSize {
+			// We have previously started writing a packet that has some space
+			// left. We can write right away.
+			remainingBytes := rlc.packetSize - rlc.writtenBytes
+			rlc.writtenBytes = rlc.packetSize
+			if uint64(len(b)) <= remainingBytes {
+				packets <- b
+				return
+			}
+			packets <- b[:remainingBytes]
+			b = b[remainingBytes:]
+		}
+		// Write the remaining data packet by packet.
+		timer := time.NewTimer(time.Until(rlc.lastWrite.Add(time.Second)))
+		for len(b) > 0 {
+			// Wait long enough
+			select {
+			case <-timer.C:
+			}
+			// Pass the data on to the writing thread
+			if uint64(len(b)) <= rlc.packetSize {
+				packets <- b
+				b = b[:0]
+				rlc.writtenBytes = uint64(len(b))
+			} else {
+				packets <- b[:rlc.packetSize]
+				b = b[rlc.packetSize:]
+				rlc.writtenBytes = rlc.packetSize
+			}
+			// Remember the last write's timestamp
+			rlc.lastWrite = time.Now()
+			timer.Reset(time.Second)
+		}
+	}()
+
+	// Write data one packet at a time.
+	for packet := range packets {
+		written, err := rlc.conn.Write(packet)
+		n += written
+		if err != nil {
+			return 0, err
+		}
+	}
+	return
+}
+
+// Close calls the underlying connection's Close method.
+func (rlc *RLConnection) Close() error {
+	return rlc.conn.Close()
+}
+
+// LocalAddr calls the underlying connection's LocalAddr method.
+func (rlc *RLConnection) LocalAddr() net.Addr {
+	return rlc.conn.LocalAddr()
+}
+
+// RemoteAddr calls the underlying connection's RemoteAddr method.
+func (rlc *RLConnection) RemoteAddr() net.Addr {
+	return rlc.conn.RemoteAddr()
+}
+
+// SetDeadline calls the underlying connection's SetDeadline method.
+func (rlc *RLConnection) SetDeadline(t time.Time) error {
+	return rlc.conn.SetDeadline(t)
+}
+
+// SetReadDeadline calls the underlying connection's SetReadDeadline method.
+func (rlc *RLConnection) SetReadDeadline(t time.Time) error {
+	return rlc.conn.SetReadDeadline(t)
+}
+
+// SetWriteDeadline calls the underlying connection's SetWriteDeadline method.
+func (rlc *RLConnection) SetWriteDeadline(t time.Time) error {
+	return rlc.conn.SetWriteDeadline(t)
+}
```

Comment:
you can avoid all this boilerplate by embedding the `net.Conn` field. Then you only need to implement `Write`.

Reply:
Doesn't that overwrite the `Write` function? I'd probably have to manually implement the private `net.conn` struct if I did that.

Reply:
when you embed a type, your type gets all of its methods. If you define a method on your type with the same name, it overwrites the embedded type's method.

Reply:
I know and that's the issue. `net.Conn` is an interface while `net.conn` is the actual implementation. If I embed `net.Conn` and overwrite `Write` I have nothing to write to since the `c.fd` field that is used in the `net.conn` implementation is not present in `net.Conn` since it is only an interface.